include/RAJA/pattern/teams.hpp

```diff
@@ -0,0 +1,676 @@
+    }
+  }
+};
+
+//collapsed cuda policies
+
+template <typename SEGMENT>
+struct LoopExecute<cuda_block_xyz_direct<2>, SEGMENT> {
+
+  template <typename BODY>
+  static RAJA_INLINE RAJA_DEVICE void exec(LaunchContext const &ctx,
+                               SEGMENT const &segment0,
+                               SEGMENT const &segment1,
+                               BODY const &body)
+  {
+    int len1 = segment1.end() - segment1.begin();
+    int len0 = segment0.end() - segment0.begin();
+    {
+      const int i = blockIdx.x; const int j = blockIdx.y;
+      body(*(segment0.begin() + i), *(segment1.begin() + j));
+    }
+  }
+};
+
+template <typename SEGMENT>
+struct LoopExecute<cuda_block_xyz_direct<3>, SEGMENT> {
+
+  template <typename BODY>
+  static RAJA_INLINE RAJA_DEVICE void exec(LaunchContext const &ctx,
+                               SEGMENT const &segment0,
+                               SEGMENT const &segment1,
+                               SEGMENT const &segment2,
+                               BODY const &body)
+  {
+    int len2 = segment2.end() - segment2.begin();
+    int len1 = segment1.end() - segment1.begin();
+    int len0 = segment0.end() - segment0.begin();
+    {
+      const int i = blockIdx.x; 
+      const int j = blockIdx.y;
+      const int k = blockIdx.z;
+      body(*(segment0.begin() + i), 
+           *(segment1.begin() + j),
+           *(segment2.begin() + k));
+    }
+  }
+};
+
+
+#endif
+
+
+template <typename POLICY_LIST, typename CONTEXT, typename SEGMENT, typename BODY>
+RAJA_HOST_DEVICE RAJA_INLINE void loop(CONTEXT const &ctx,
+                           SEGMENT const &segment,
+                           BODY const &body)
+{
+#ifdef __CUDA_ARCH__
+  LoopExecute<typename POLICY_LIST::device_policy_t,
+              SEGMENT>::exec(ctx, segment, body);
+#else
+  switch (ctx.exec_place)
+  {
+     case HOST_THREADS: LoopExecute<typename POLICY_LIST::host_threads_policy_t,
+                           SEGMENT>::exec(ctx, segment, body); break;
+
+     case HOST: LoopExecute<typename POLICY_LIST::host_policy_t,
+                            SEGMENT>::exec(ctx, segment, body); break;
+  }
+#endif
+}
+
+template <typename POLICY_LIST, typename CONTEXT, typename SEGMENT, typename BODY>
+RAJA_HOST_DEVICE RAJA_INLINE void loop(CONTEXT const &ctx,
+                           SEGMENT const &segment0,
+                           SEGMENT const &segment1,
+                           BODY const &body)
+{
+#ifdef __CUDA_ARCH__
+  LoopExecute<typename POLICY_LIST::device_policy_t,
+              SEGMENT>::exec(ctx, segment0, segment1, body);
+#else
+  switch (ctx.exec_place)
+  {
+     case HOST_THREADS: LoopExecute<typename POLICY_LIST::host_threads_policy_t,
+                                    SEGMENT>::exec(ctx, segment0, segment1, body); break;
+
+     case HOST: LoopExecute<typename POLICY_LIST::host_policy_t,
+                            SEGMENT>::exec(ctx, segment0, segment1, body); break;
```

Comment:
Folks, I added a switch statement to switch between host sequential and threads. I'm not certain on the overhead here.... but this is one way to add the omp cpu threads backend.